### ISSUE TITLE
First test results

### ISSUE THREAD

**Issue by YouCanNotBeSerious** [RobTillaart/CHT8310#3]
Referencing - https://github.com/RobTillaart/CHT8305/issues/14

I have tested on new hardware, and found that the temperature readings align within tolerance to reality, but the humidity readings are unstable.  

I will mention that the humidity readings are less important for my application, so your work has enabled my project to continue, and I imagine others who find this library will be thankful as well.  
With that in mind, I am happy to test more if it helps, otherwise it should be noted that in limited testing this lib basically works for temperature, but humidity needs improvement.

**Comment by YouCanNotBeSerious** [comment]
Data - temperature in degC settles well to room temperature after being hit with heat gun, and readings are consistent with no wild values, but humidity is unstable.

<img width="1280" alt="Screen Shot 2024-02-14 at 6 29 00 PM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/f7ea1ca1-b3e9-4b6a-bea8-146f71aee0c0">

<img width="1280" alt="Screen Shot 2024-02-14 at 6 29 10 PM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/cdaaf23f-66d9-46d5-9b42-380c854b4881">

**Comment by RobTillaart** [comment]
Good morning,
Do you think the "humidity horror" is a library bug / problem or a device issue?

**Comment by YouCanNotBeSerious** [comment]
I have done very limited testing - and will be continuing but with the same hardware and code the temperature readings are good so my instinct is library.  

I am by no means a coder, more a hacker, but i notice the pattern is humidity seems to go positive and negative wildly, could it be something running off the end of a limit (it's not quite pos and neg 127 but close....)  since I believe you had to dive deep into registers and things compared to the 8305 version of the chip?

I am happy to run or gather whatever debug would help  :-)

**Comment by RobTillaart** [comment]
I will dive into the datasheets again, if my agenda permits to free some time..
To be continued

**Comment by RobTillaart** [comment]
The humidity register has an overflow bit, that needs to be handled.
Will set humidity to 100% and return an error (read function).

Develop branch will be created asap.

**Comment by RobTillaart** [comment]
Develop branch created with handling of the overflow bit.
if set the read() function sets humidity to 100% and returns **CHT8310_ERROR_HUMIDITY**

Please test if the humidity graph looks (a bit) better.

What is the read/update frequency in your code?

**Comment by RobTillaart** [comment]
- Added setting of the limit registers to trigger the ALERT pin
- Added **uint16_t getStatusRegister()**
- added minimal example for ALERT.
- Update readme.md

Please test if this new functionality works as expected.

**Comment by RobTillaart** [comment]
Pull request #4 created,

**Comment by RobTillaart** [comment]
- added **softwareReset()** experimental
- added **setConvertRate(uint8_t rate = 4)**  4 is one sample per second.
- added **getConvertRate()**

With the convertRate you might improve the humidity.

Looked again at your graph of humidity and it looked like the overflow negated a value.
I'm wondering if that was the last read value or a new value. 
Datasheet doesn't say anything about these bits.

**Comment by RobTillaart** [comment]
- added **oneShotConversion()**

so far all the simple code. 
Still open is the configuration register, need to read the datasheet carefully to see what it can do.

**Comment by YouCanNotBeSerious** [comment]
Hi - should I be testing something specific?  Updated to the dev branch, and the humidity limits as expected at 100 but readings are still wildly varying.  

This screenshot with updated lib, no other changes.

<img width="1368" alt="Screen Shot 2024-02-15 at 10 07 28 AM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/5013be79-e0f3-4aac-8dcf-e93c4bcf9146">

This one is with CHT.setConvertRate(4); - both temp and humidity don't move now. Tried setConvertRate with 1,4,8,16 and the values for both temp and hum stay static.

<img width="797" alt="Screen Shot 2024-02-15 at 10 16 14 AM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/38a53f91-a01f-4fbe-9ac6-5204eb3a7037">

**Comment by RobTillaart** [comment]
You could try other convert rates that might stabilize readings.
Maybe test if the temp limits work using the alert pin.
Had no time to think over and write a good test case.

**Comment by YouCanNotBeSerious** [comment]
I just edited my previous comment....  In a selfish way, my problem is solved as I can use temperature with the first version of the lib or by not specifying setConvert rate, and my hardware now works instead of trying to SMT replace the 8310 with 8305 :-).    So I am not pushing for the humidity to work, you have solved my issue and I am grateful!

But I will easily test any other changes you suggest, or if you have a way for me to post the hardware cheaply you are welcome to keep it - it is this, with ESP32-C3 replacing the original micro:

<img width="554" alt="Screen Shot 2024-02-15 at 10 21 22 AM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/21328a63-f835-498a-881b-407f61166bd6">

**Comment by RobTillaart** [comment]
>  Tried setConvertRate with 1,4,8,16 and the values for both temp and hum stay static.

mmm lets think...

ConvertRate should be 0 .. 7  See readme.md

|  value |  measurement  |  frequency  |  notes  |
|:------:|:-------------:|:-----------:|:-------:|
|   0    |     120 s     |  1/120 Hz   |
|   1    |      60 s     |   1/60 Hz   |
|   2    |      10 s     |    0.1 Hz   |
|   3    |       5 s     |    0.2 Hz   |
|   4    |       1 s     |      1 Hz   |  default
|   5    |     500 ms    |      2 Hz   |
|   6    |     250 ms    |      4 Hz   |
|   7    |     125 ms    |      8 Hz   |

When > 7 it  is set to 7 which is fastest and thus less precise.

Can you print the return value of **getConvertRate()** after power up?
Can you print the return value of **readRegister(4)** after power up?

**Comment by RobTillaart** [comment]
>  if you have a way for me to post the hardware cheaply you are welcome to keep it

Thanks for the offer, appreciated.
However recent changes (customs) here would easily put 20 euro (or more) import tax, even on a gift.

**Comment by YouCanNotBeSerious** [comment]
> > if you have a way for me to post the hardware cheaply you are welcome to keep it
> 
> Thanks for the offer, appreciated. However recent changes (customs) here would easily put 20 euro (or more) import tax, even on a gift.

I imagined something like that :-(

>Can you print the return value of getConvertRate() after power up?
>Can you print the return value of readRegister(4) after power up?

***startup***
get convertrate() gives: 4
readRegister(4) gives: 1024
100.00	24.72
100.00	24.72
100.00	24.72


BUT putting setConvertRate(2) in setup() and rerunning, we get:
***startup***
get convertrate() gives: 0
readRegister(4) gives: 0
100.00	24.84
100.00	24.84
100.00	24.84


Same with setConvertRate 0,2,4 i.e. trying to set it at all.

**Comment by RobTillaart** [comment]
> BUT putting setConvertRate(2) in setup() and rerunning, we get:
startup
get convertrate() gives: 0
readRegister(4) gives: 0

Good test, 
I need to check the register write (later today) as it should return the value set.
As the default value is 4 the register read works as intended.

**Comment by RobTillaart** [comment]
Think I found the problem, looks like I swapped the bytes in writeRegister.

Develop branch has a patch, can you test.

**Comment by YouCanNotBeSerious** [comment]
Repeated tests with latest from dev branch - 

default, not setting conversion rate (should be 4):
***startup***
get convertrate() gives: 4
readRegister(4) gives: 1024
100.00	23.97
100.00	23.97
100.00	23.97
100.00	24.00


Setting to 2:
***startup***
get convertrate() gives: 2
readRegister(4) gives: 512
93.75	22.81
93.75	22.81
100.00	22.94
100.00	22.94
100.00	22.94

Setting to 0:
***startup***
get convertrate() gives: 0
readRegister(4) gives: 0
100.00	23.53
100.00	23.53
100.00	23.53
100.00	23.53
100.00	23.53
100.00	23.53
100.00	23.53

So the register fix seems good, but the effect on temperature and humidity readings is that temp is still good, humidity railing most of the time to 100.

**Comment by RobTillaart** [comment]
So an important step forward.
I read the config register has a bit to switch a heater on. It might be possible that the humidity sensor has some moisture build up. The heater could evaporate that. .
But not seen how long a heat session should take.

**Comment by YouCanNotBeSerious** [comment]
Gave it a good blast with a SMD rework hot air gun with thermocouple on the PCB near the sensor reading ~100C for ~20s :-)     Temperature looks good at least !

The humidity does react to blowing air across the sensor, but the pattern seems to be it settles at or near 100 once back at stable room temp and humidity.

<img width="1644" alt="Screen Shot 2024-02-16 at 10 17 36 AM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/c2ace9b5-a5e5-4fd9-9f57-7510207950ca">

**Comment by RobTillaart** [comment]
Can you print(readRegister(1)). In the loop?
I'm interested if the raw register value of the humidity does jump too.
If not it is a library problem.

**Comment by RobTillaart** [comment]
Another option to consider is removing and ignoring the overflow bit. 
There is a chance that the remaining bits are not bad...
Need a dozen 'raw humidities' for that.

**Comment by YouCanNotBeSerious** [comment]
Changed to print raw value from readRegister(1) and getHumidity() since temperature values are not in question here:

### With convertRate(4):

***startup***
get convertrate() gives: 4
readRegister(4) gives: 1024
12397	28.90
12580	22.65
12718	28.90
12823	22.65
12879	22.65
12902	22.65
12902	22.65
13052	66.41
13304	22.65
13823	22.65
14039	66.41
13861	66.41
13466	22.65
13117	22.65
12861	28.90
12657	28.90
12533	100.00
12504	28.90
12455	28.90
12366	100.00
12388	28.90
12384	100.00
12331	100.00
12270	100.00
12270	100.00
12282	100.00
12428	100.00
12533	100.00
12479	100.00
12549	100.00
12544	28.90
12731	28.90
13060	28.90
13081	28.90
12907	28.90
12667	100.00
12516	28.90
12427	28.90
12346	28.90
12286	100.00
12414	28.90
12414	28.90
12459	100.00
12420	28.90
12424	100.00
12356	28.90
12329	100.00
12414	28.90
12544	28.90
12713	28.90
12914	28.90
13119	22.65
13131	22.65
13004	28.90
12891	28.90
13025	28.90
13284	22.65
13291	28.90
13291	28.90
13139	28.90
12992	22.65
13048	22.65
13209	22.65
13156	22.65

### With convertRate(0):
***startup***
get convertrate() gives: 0
readRegister(4) gives: 0
12246	100.00
12246	100.00
12246	100.00
12246	100.00
cut - did not change for at least 100 more prints


### With convertRate(2):
***startup***
get convertrate() gives: 2
readRegister(4) gives: 512
12063	2.34
12063	2.34
12063	2.34
12063	2.34
12063	2.34
12063	2.34
12063	2.34
12063	2.34
12063	2.34
12063	2.34
13544	58.59
13544	58.59
13544	58.59
13544	58.59
13544	58.59
13544	58.59
13544	58.59
13544	58.59
13544	58.59
13544	58.59
13544	58.59
14416	58.59
14416	58.59
14416	58.59
14416	58.59
14416	58.59
14416	58.59
12987	85.94
12987	85.94
12987	85.94
12987	85.94
12987	85.94
12987	85.94
12987	85.94
12987	85.94

**Comment by RobTillaart** [comment]
Thanks,, the raw data seems ok.
So the library seems to either corrupt the data
or fail to read raw data correctly
Or ...

Investigate tomorrow

**Comment by RobTillaart** [comment]
Some raw register reads put in a graph (divide by 327.67 as it should

![image](https://github.com/RobTillaart/CHT8310/assets/462844/7008c284-03eb-4390-96ea-17ceabb70cd1)

So back to drawing board, why does the library fail to handle humidity?

**Comment by RobTillaart** [comment]
Pushed a new version 0.2.0 with an explicit read of the humidity register in **read()**.
In the previous version I tried to read 2 register in one call and this might be incorrect.

Can you please run the new test example included and post its output (10 lines or so).
Thanks!

**Comment by YouCanNotBeSerious** [comment]
New test example of 0.2.0 output below - I only changed the wire.begin line for my SCL/SDA pins, no other changes.

This looks good to me - let it run for several hundred readings in a large room, still air, no overt humidity or temperature changes and it looks a) believable compared to seperate humidity and temperature monitoring and b) stable over time.

Second set of values is one calibrated human blowing over the sensor to change humidity and temp :-). Again looks believable to have both rise then settle back to near previous values.

**I think this is it - awesome! :-)**

TEST1
205107	0	16078	49.07	5736	22.41
206108	0	16061	49.02	5744	22.44
207108	0	16045	48.97	5744	22.44
208109	0	16025	48.91	5744	22.44
209109	0	16030	48.92	5736	22.41
210110	0	16027	48.91	5744	22.44
211110	0	16027	48.91	5744	22.44
212111	0	16016	48.88	5744	22.44
213111	0	16018	48.88	5744	22.44
214112	0	16000	48.83	5744	22.44
215112	0	15995	48.81	5744	22.44

<img width="912" alt="Screen Shot 2024-02-17 at 10 57 35 AM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/56f48801-6ce2-4f4b-8527-b0697d5f6d65">




TEST2
282146	0	16522	50.42	5760	22.50
283146	0	16641	50.79	5768	22.53
284147	0	18963	57.87	5808	22.69
285147	0	20153	61.50	5936	23.19
286148	0	22838	69.70	6088	23.78
287148	0	25295	77.20	5968	23.31
288149	0	26702	81.49	6328	24.72
289149	0	28271	86.28	6360	24.84
290150	0	29440	89.85	6240	24.37
291150	0	29440	89.85	6240	24.37
292151	0	30320	92.53	6176	24.12
293151	0	30696	93.68	6104	23.84
294152	0	28808	87.92	6064	23.69
295152	0	27071	82.62	6024	23.53
296153	0	25285	77.17	6008	23.47
297153	0	23818	72.69	5992	23.41
298154	0	22432	68.46	5984	23.37
299154	0	21177	64.63	5976	23.34

<img width="912" alt="Screen Shot 2024-02-17 at 10 56 13 AM" src="https://github.com/RobTillaart/CHT8310/assets/41044464/ad7b39ad-de21-443c-b8b2-bb2481dc4aef">

**Comment by RobTillaart** [comment]
Finally, this is what it should look like. 

The main change is reading the registers one by one. Apparently when trying to read two registers in one call the second pair of values is not taken from the expected second register. Reading "adjacent" registers in one call is common practice but seems not supported by this device. 
This hypothesis is in line with the tests done.

Next action for me is to cleanup the code, check documentation and release this working version.

Then writing examples for the other functions like the alert. These should be verified/tested some day too.

Last but not least, an usable set of functions for the configuration should be made. 

Given the current insights these actions should not give too much trouble.

To be continued

**Comment by RobTillaart** [comment]
Btw thanks for your patience and test efforts!

**Comment by YouCanNotBeSerious** [comment]
> Btw thanks for your patience and test efforts!

Welcome - I can hack together arduino code but really am a hardware person.  Without library devs like yourself, none of my projects would ever get off the ground!

As far as closing the issue - I imagine you do this when the release comes?  I can confirm on the final release if you like, and then we close?

**Comment by RobTillaart** [comment]
>  As far as closing the issue - I imagine you do this when the release comes? I can confirm on the final release if you like, and then we close?

Yes that would be the process. With the current issue we got the library working to a useful level.
Next steps will cover and verify remaining functionality, resulting in subsequent releases.
So that would be new issues to have "one issue per bug / feature"

Would be great if you could run through the examples which I'm preparing now to see if there are artefacts.

Next update expected within an hour or so.

**Comment by RobTillaart** [comment]
@YouCanNotBeSerious 
Beta 0.2.0 pushed to develop branch,
if you have time please verify the examples.

**Comment by YouCanNotBeSerious** [comment]
Re ran the exact same CH8310_test code as previously before flashing the latest, to get the room current temp and humidity:

### CHT8310_test_prev_dev_release 

2097	0	17744	54.15	5768	22.53
3097	0	17690	53.99	5768	22.53
4097	0	17584	53.66	5768	22.53
5097	0	17497	53.40	5768	22.53
6097	0	17483	53.36	5768	22.53



Updated to latest dev 0.2.0 from this point

### CHT8310_test:
1097	0	16961	51.76	5944	23.22
2097	0	16961	51.76	5944	23.22
3097	0	16905	51.59	5944	23.22
4097	0	16869	51.48	5944	23.22
5097	0	16884	51.53	5944	23.22
6097	0	16927	51.66	5944	23.22
7097	0	16877	51.51	5952	23.25


### CHT8310_performance:
200000
READ:	2
READ T:	329
READ H:	2
M:8215:	328

250000
READ:	2
READ T:	283
READ H:	2
M:8215:	282

### CHT_setConvertRate:
Not actually very noisy - flicks around in the decimal places a bit but not much from reading to reading. Seems like value is cached and updates at about 1s rate even though the serial output is faster.  This data has large humidity change from blowing on the sensor briefly. Watching the data scroll past for a few minutes it was monotonic i.e. up with breath then trending down, no readings went back up randomly.  When sensor is left undisturbed after a few minutes the readings stabilise within about 2 RH and very stable on temperature, within about 0.1C mostly.

194888	66.58	24.66
194889	66.58	24.66
194889	66.58	24.66
194889	66.58	24.66
194889	66.58	24.66
194900	66.58	24.66
194900	66.58	24.66
194900	66.58	24.66
194900	66.58	24.66
194900	66.58	24.66
194901	66.58	24.66
194901	66.58	24.66
194911	66.58	24.66
194911	66.58	24.66
194912	66.58	24.66
194912	66.58	24.66
194912	66.58	24.66
194912	66.58	24.66
194912	66.58	24.66
194924	73.71	24.78
194924	73.71	24.78
194924	73.71	24.78
194925	73.71	24.78
194925	73.71	24.78
194925	73.71	24.78
194925	73.71	24.78
194936	73.71	24.78
194936	73.71	24.78
194936	73.71	24.78
194936	73.71	24.78
194937	73.71	24.78
194937	73.71	24.78
194937	73.71	24.78
194947	73.71	24.78
194948	73.71	24.78
194948	73.71	24.78
194948	73.71	24.78
194948	73.71	24.78
194948	73.71	24.78
194949	73.71	24.78
194959	73.71	24.78
194959	73.71	24.78
194959	73.71	24.78
194960	73.71	24.78
194960	73.71	24.78

###device_reset

Device reset success: 	4

Note header comments in a few examples still have 8305 still referenced in the URL, line 5.

Do you want longer captures or anything else run?

**Comment by RobTillaart** [comment]
Thanks!
OK, examples need a minor edit 

Longer output, only from the performance to see how the read "improves" at different I2C clocks.

**Comment by RobTillaart** [comment]
URL and name in examples updated (did repo search for cht8305 string)

**Comment by RobTillaart** [comment]
Had a look at this output
```
CHT8310_performance:
200000
READ: 2
READ T: 329
READ H: 2
M:8215: 328

250000
READ: 2
READ T: 283
READ H: 2
M:8215: 282
```

The manufacturers ID of 8215 is not what is expected from the datasheet. It is no showstopper but at least remarkable.
Furthermore the READ and READ H of 2 micros needs some investigation.

In 30 minutes a new update is expected.

**Comment by RobTillaart** [comment]
Pushed patch for performance test.

**Comment by YouCanNotBeSerious** [comment]
Hmmm re-run performance test with latest 0.2.0 dev, still lots of single digit READs:


50000
READ:	2165
READ T:	1060
READ H:	3
M:8215:	1056

100000
READ:	3
READ T:	582
READ H:	3
M:8215:	583

150000
READ:	3
READ T:	428
READ H:	3
M:8215:	428

200000
READ:	3
READ T:	327
READ H:	3
M:8215:	328

250000
READ:	3
READ T:	281
READ H:	3
M:8215:	282

300000
READ:	3
READ T:	246
READ H:	3
M:8215:	246

350000
READ:	3
READ T:	225
READ H:	3
M:8215:	226

400000
READ:	3
READ T:	209
READ H:	3
M:8215:	210

450000
READ:	3
READ T:	195
READ H:	3
M:8215:	196

500000
READ:	3
READ T:	186
READ H:	3
M:8215:	187

550000
READ:	3
READ T:	177
READ H:	3
M:8215:	177

600000
READ:	3
READ T:	173
READ H:	3
M:8215:	173

**Comment by RobTillaart** [comment]
Can you change the delay in the test function to 1000 millis?

**Comment by YouCanNotBeSerious** [comment]
Done:

```
void testPerformance(uint32_t speed)
{
  start = micros();
  CHT.read();
...
  **delay(1000);**

  start = micros();
  CHT.readTemperature();
...
  **delay(1000);**

  start = micros();
  CHT.read();
...
  **delay(1000);**
  
  start = micros();
  uint16_t m = CHT.getManufacturer();
...
  **delay(1000);**
}
```

50000
READ:	2165
READ T:	1060
READ H:	2116
M:8215:	1051

100000
READ:	1164
READ T:	582
READ H:	1168
M:8215:	578

150000
READ:	855
READ T:	428
READ H:	860
M:8215:	424

200000
READ:	654
READ T:	327
READ H:	658
M:8215:	323

250000
READ:	562
READ T:	281
READ H:	566
M:8215:	277

300000
READ:	491
READ T:	246
READ H:	496
M:8215:	242

350000
READ:	450
READ T:	225
READ H:	454
M:8215:	221

400000
READ:	418
READ T:	209
READ H:	422
M:8215:	205

450000
READ:	390
READ T:	195
READ H:	394
M:8215:	191

500000
READ:	372
READ T:	186
READ H:	376
M:8215:	182

550000
READ:	353
READ T:	177
READ H:	358
M:8215:	173

600000
READ:	345
READ T:	173
READ H:	349
M:8215:	169

**Comment by RobTillaart** [comment]
Looks more like real fetching times
Think that the sensor is busy when you request too fast, causing a failure on I2C responsiveness.
For now these are good as indicative performance figures.
What processor?  ESP32-C3?

**Comment by RobTillaart** [comment]
Stil wondering why readhumity() is slow.
Next release issue

**Comment by YouCanNotBeSerious** [comment]
> What processor? ESP32-C3?

Correct!  ESP8685-WROOM-06 is an ESP32-C3 module pin compatible with the Tuya proprietary module in a lot of wifi home automation things they make, and a LOT of other companies white label those products and sell.  

With the module replaced I can use for example Tasmota, esphome, or my preferred HomeSpan arduino library to have nice mass produced sensors, switches, other home automation products that are native HomeKit, or MQTT for Home Asstant or whatever.

Point is, these temperature sensors are similar but cheaper than other versions, except they use CHT8310...whcih now thanks to your library, is not a problem :-)

**Comment by RobTillaart** [comment]
> Stil wondering why readhumity() is slow.
Next release issue

found the cause ==> copy/paste bug in the performance sketch (stupid me).
A fix will be posted right away.

**Comment by RobTillaart** [comment]
Posted

**Comment by YouCanNotBeSerious** [comment]
Looks more believable!

50000
READ:	2166
READ T:	1060
READ H:	1058
M:8215:	1052

100000
READ:	1163
READ T:	582
READ H:	582
M:8215:	578

150000
READ:	855
READ T:	427
READ H:	427
M:8215:	423

200000
READ:	653
READ T:	327
READ H:	327
M:8215:	323

250000
READ:	561
READ T:	281
READ H:	281
M:8215:	277

300000
READ:	491
READ T:	245
READ H:	245
M:8215:	241

350000
READ:	449
READ T:	225
READ H:	225
M:8215:	221

400000
READ:	417
READ T:	209
READ H:	209
M:8215:	204

450000
READ:	389
READ T:	195
READ H:	195
M:8215:	191

500000
READ:	371
READ T:	186
READ H:	186
M:8215:	181

550000
READ:	353
READ T:	176
READ H:	177
M:8215:	172

600000
READ:	344
READ T:	172
READ H:	172
M:8215:	168

**Comment by RobTillaart** [comment]
Thanks for these new numbers.
I will update the readme.md tomorrow and release 0.2.0

**Comment by RobTillaart** [comment]
Merged and released as 0.2.0
